checks/datadog.py

```diff
@@ -25,9 +25,11 @@ def partition(s, sep):
 
 
 def point_sorter(p):
-    # Sort and group by timestamp, metric name, host_name, device_name
-    return (p[1], p[0], p[3].get('host_name', None), p[3].get('device_name', None))
-
+    # Sort and group by timestamp, metric name, host_name, device_name, (tags or attributes)
```

Comment:
reason why I did it this way.

As per [Our docs](https://docs.datadoghq.com/agent/faq/dogstream/#datadog-canonical-log-format), the conical format is
`metric unix_timestamp value [attribute1=v1 attributes2=v2 ...]`

And for ParserFunction-parsed logs, attributes contain `tags`.
```
    metric_attributes = {
        'tags': tags,
        'metric_type': 'gauge',
    }
    return (metric_name, date, metric_value, metric_attributes)
```

So ideally in canonical format, metric tags should be sent like below - which is why I look for `tags`
`mymetric 1531900000 99 tags=env:prod,app:web metric_type=gauge`

BUT, most of the time in practice (and from an [old KB content](https://help.datadoghq.com/hc/en-us/articles/210645523-How-to-collect-metrics-with-the-Agent-s-built-in-Log-Parser)), the tags are sent as follows - also because the example we usually provide looks like below; in this case there are no `tags` so we assume that the `attribs` could contain the tags.
`applications.function.runtime_seconds 1464462187 24 application=myapp code_author=gus`

Reason why NOT to use attribs all the time is because we have to assume that the tags (if specified) are ordered - which isn't always the case. Example below where using `attribs` only doesn't work
1_ Using canonical format, this would look like 2 unique metrics and will not be aggregated.
```
mymetric 1531900000 99 tags=env:prod,app:web metric_type=gauge
mymetric 1531900000 99 tags=app:web,env:prod metric_type=gauge
```
2_ using parser function, same thing. If the tags were submitted in a different order, they'll not be aggregated; we can't assume that `tags` are always sorted when returned.
```
    tags = extras.split(',')
    metric_attributes = {
        'tags': tags,
        'metric_type': 'gauge',
    }
```

Reply:
Thanks a lot for that great explication !

Should we not also sort `p[3]` if `tags` is not set ? Right now we're only sorting `tags`.

Reply:
Apparently, theres's no need. Per my testing, we don't need to sort *dict*; only needed to do it for *list*. Took a while for me to figure this out: ¯\_(ツ)_/¯

- `attribs` returned is a *list* if `tags` were found 
- `attribs` returned is a *dict* if from `p[3]`.